lib/semantics/semantics.cc

```diff
@@ -168,6 +298,24 @@ Scope &SemanticsContext::FindScope(parser::CharBlock source) {
   }
 }
 
+void SemanticsContext::PushConstruct(const ConstructNode &construct) {
```

Comment:
I'm actually surprised that C++ implicitly converts the pointers at the various call sites into a `ConstructNode` that can be passed by reference.

This function might be more clear if it used move semantics.

Reply:
I want to make sure that I understand your suggestion.  Are you saying that, for example, the code for the ```Pre``` function for a ```DoConstruct``` should look like this?
```
  bool Pre(const parser::DoConstruct &doConstruct) {
    ConstructNode doNode{std::move(&doConstruct)};
    context_.PushConstruct(doNode);
    Enter(doConstruct);
    return true;
  }
```

Reply:
No, that doesn't work.  I meant something like `context_.PushConstruct(ConstructNode{doConstruct});` in `Pre()` with the definition of `PushConstruct` accepting a `ConstructNode &&construct` argument.

Reply:
Thanks, Peter.  I'll change the ```Pre()``` functions.

Reply:
Take a look at the version in PR #708

Reply:
In Brian's code, ```PushConstruct()``` is a template that updates the data member ```constructStack_```.  But since ```constructStack_``` is of a non-template type, I don't see the advantage of implementing ```PushConstruct()``` as a template.

Why make it a template?

Reply:
The template suggestion doesn't have anything to do with `PushConstruct` or `constructStack_` being templated. This PR has roughly 10 `Pre()` and `Post()` routines manually defined, functionally they all do exactly the same thing. Instead you could use a templated routine each for `Pre()` and `Post()` .

Reply:
Please disregard my comment. I misunderstood things.

Reply:
Thanks, Tim.  I'll change my code to match Brian's.